### ISSUE TITLE
MSecurityManager blocks write access to jacoco.exec during Maven build (v.1.0.5)

### ISSUE THREAD

**Issue by JavaJimFl** [EvoSuite/evosuite#120]
### Context
*Please provide below a detailed introduction to the issue itself, and describe what you were doing when the issue happened. Or, what do you want to achieve?*

I'm trying to generate JaCoCo code coverage metrics during a Maven build, but the sandbox won't allow write access to the JaCoCo data file (target/jacoco.exec).  The project is JMockit-enabled using their agent, but the production code is instrumented offline by JaCoCo.  I have tried setting `sandbox=false` and `sandbox_mode=OFF` in the evosuite.properties file.  

Note: Even though I'm using the EvoSuite Maven plugin, the runtime appears to honor the properties file while ignoring the `extraArgs` element of the plugin configuration.  I am successfully using it to tweak other aspects of the test generation, such as the `test_scaffolding`, `test_dir`,  and  `test_naming_strategy`properties.  I think this may be related to issue [#66](https://github.com/EvoSuite/evosuite/issues/66).

The only thing that has worked for me is creating a temporary build where I  commented out the body of the `MSecurityManager.allowPermission()` method, having it simply return `true`.

### Steps to Reproduce
*Please break down here below all the needed steps to reproduce the issue.*
[If possible, please upload an example of the project you are generating tests for.]

1.  
2.
3.

### EvoSuite Arguments
*Please provide the whole EvoSuite commmand you executed (if relevant)*
`mvn clean install`

### Current Result
*Please describe here below the current result you got (if relevant)*
[if relevant, include a screenshot]
```java
java.lang.SecurityException: Security manager blocks ("java.io.FilePermission" "C:\MyProject\target\jacoco.exec" "write")
java.lang.Thread.getStackTrace(Thread.java:1556)
org.evosuite.runtime.sandbox.MSecurityManager.checkPermission(MSecurityManager.java:434)
java.lang.SecurityManager.checkWrite(SecurityManager.java:979)
java.io.FileOutputStream.<init>(FileOutputStream.java:200)
org.jacoco.agent.rt.internal_8ff85ea.output.FileOutput.openFile(FileOutput.java:67)
org.jacoco.agent.rt.internal_8ff85ea.output.FileOutput.startup(FileOutput.java:49)
org.jacoco.agent.rt.internal_8ff85ea.Agent.startup(Agent.java:122)
org.jacoco.agent.rt.internal_8ff85ea.Agent.getInstance(Agent.java:50)
org.jacoco.agent.rt.internal_8ff85ea.Offline.<clinit>(Offline.java:31)
com.fple.fim.econfirm.EConfirmApiMethod.$jacocoInit(EConfirmApiMethod.java)
com.fple.fim.econfirm.EConfirmApiMethod.<clinit>(EConfirmApiMethod.java)
java.lang.Class.forName0(Native Method)
java.lang.Class.forName(Class.java:348)
org.evosuite.runtime.classhandling.ClassStateSupport.loadClasses(ClassStateSupport.java:136)
org.evosuite.runtime.classhandling.ClassStateSupport.initializeClasses(ClassStateSupport.java:62)
com.fple.fim.econfirm.EConfirmApiMethodEsTest.initializeClasses(EConfirmApiMethodEsTest.java:88)
com.fple.fim.econfirm.EConfirmApiMethodEsTest.initEvoSuiteFramework(EConfirmApiMethodEsTest.java:44)
sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.lang.reflect.Method.invoke(Method.java:498)
org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
mockit.integration.junit4.internal.JUnit4TestRunnerDecorator.executeClassMethod(JUnit4TestRunnerDecorator.java:92)
mockit.integration.junit4.internal.JUnit4TestRunnerDecorator.invokeExplosively(JUnit4TestRunnerDecorator.java:30)
mockit.integration.junit4.internal.MockFrameworkMethod.invokeExplosively(MockFrameworkMethod.java:37)
org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java)
org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
org.junit.runners.ParentRunner.run(ParentRunner.java:363)
org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)

	at org.evosuite.runtime.sandbox.MSecurityManager.checkPermission(MSecurityManager.java:452)
	at java.lang.SecurityManager.checkWrite(SecurityManager.java:979)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:200)
	at org.jacoco.agent.rt.internal_8ff85ea.output.FileOutput.openFile(FileOutput.java:67)
	at org.jacoco.agent.rt.internal_8ff85ea.output.FileOutput.startup(FileOutput.java:49)
	at org.jacoco.agent.rt.internal_8ff85ea.Agent.startup(Agent.java:122)
	at org.jacoco.agent.rt.internal_8ff85ea.Agent.getInstance(Agent.java:50)
	at org.jacoco.agent.rt.internal_8ff85ea.Offline.<clinit>(Offline.java:31)
	at com.fple.fim.econfirm.EConfirmApiMethod.$jacocoInit(EConfirmApiMethod.java)
	at com.fple.fim.econfirm.EConfirmApiMethod.<clinit>(EConfirmApiMethod.java)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.evosuite.runtime.classhandling.ClassStateSupport.loadClasses(ClassStateSupport.java:136)
	at org.evosuite.runtime.classhandling.ClassStateSupport.initializeClasses(ClassStateSupport.java:62)
	at com.fple.fim.econfirm.EConfirmApiMethodEsTest.initializeClasses(EConfirmApiMethodEsTest.java:88)
	at com.fple.fim.econfirm.EConfirmApiMethodEsTest.initEvoSuiteFramework(EConfirmApiMethodEsTest.java:44)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at mockit.integration.junit4.internal.JUnit4TestRunnerDecorator.executeClassMethod(JUnit4TestRunnerDecorator.java:92)
	at mockit.integration.junit4.internal.JUnit4TestRunnerDecorator.invokeExplosively(JUnit4TestRunnerDecorator.java:30)
	at mockit.integration.junit4.internal.MockFrameworkMethod.invokeExplosively(MockFrameworkMethod.java:37)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
```

 
 
### Expected result
*Please describe here below what should be the expected behaviour (if relevant)*

I would expect that the JaCcCo data file is writeable during the Maven build process,

### Additional info
*Please add any information of interest here below*

**Comment by gofraser** [comment]
Thanks for reporting this. Looks like we will need to add an exception to the security manager rules to allow Jacoco to write to files. Just to confirm, Jacoco works alright if you set sandbox=off, right?

**Comment by JavaJimFl** [comment]
I could find no combination of sandbox-related properties that allowed Maven/Jacoco access to the Jacoco data file.  I had to create a custom build that hard coded the return from `MSecurityManager.allowPermission()` (true) in order to capture code coverage information.

**Comment by JavaJimFl** [comment]
I just noticed I wrote sandbox=none in the original post.  It should be sandbox=off.

**Comment by realnikunj** [comment]
I am facing the exact same issue with Maven/Jacoco.
Is there a workaround for this?

**Comment by kpepper** [comment]
So am I - for ant/Evosuite/Jacoco.

**Comment by realnikunj** [comment]
So, for now - do we have to just create a custom build from Evosuite source code as suggested by @JavaJimFl above ?

**Comment by gofraser** [comment]
See http://www.evosuite.org/documentation/measuring-code-coverage/ for details and examples - this should work if following the instructions.

**Comment by JavaJimFl** [comment]
It's been a long time since I created this ticket, but I tried both online and offline instrumentation.  I recall that I could either have code coverage results or working JMockit tests, but not both.   I opted for using online instrumentation so my hand-written JMockit-based tests worked and received the error above.

Since EvoSuite, JMockit, and PIT all use instrumentation agents,  I ran into even more problems trying to use all three frameworks together and opened #131 to document my findings.

What I'm trying to say is that offline JaCoCo instrumentation may be the workaround to this particular problem, the larger issue is getting these three bytecode manipulation frameworks (and potentially others in the future) to work together in the same Maven build.
